envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RBACFilterFactory.kt

```diff
@@ -52,95 +53,146 @@ class RBACFilterFactory(
+    private fun getIncomingEndpointPolicies(
+        serviceName: String,
+        incomingPermissions: Incoming,
+        snapshot: GlobalSnapshot,
+        roles: List<Role>
+    ): List<EndpointWithPolicy> {
+        val principalCache = mutableMapOf<ClientWithSelector, List<Principal>>()
+        return incomingPermissions.endpoints
+            .mapNotNull { incomingEndpoint ->
+                if (incomingEndpoint.clients.isEmpty()) {
+                    logger.warn("An incoming endpoint definition for $serviceName does not have any clients defined." +
                         "It means that no one will be able to contact that endpoint.")
-                return@forEach
+                    return@mapNotNull null
+                }
+
+                val clientsWithSelectors = resolveClientsWithSelectors(incomingEndpoint, roles)
+                val principals = clientsWithSelectors.flatMap { client ->
+                    principalCache.computeIfAbsent(client) { mapClientWithSelectorToPrincipals(it, snapshot) }
+                }.toSet()
+
+                if (principals.isNotEmpty()) {
+                    val policy = Policy.newBuilder().addAllPrincipals(principals)
+                    val combinedPermissions = rBACFilterPermissions.createCombinedPermissions(incomingEndpoint)
+                    policy.addPermissions(combinedPermissions)
+                    EndpointWithPolicy(incomingEndpoint, policy)
+                } else {
+                    null
+                }
             }
+    }
 
-            val clientsWithSelectors = resolveClientsWithSelectors(incomingEndpoint, incomingPermissions.roles)
-            val principals = clientsWithSelectors.flatMap { mapClientWithSelectorToPrincipals(it, snapshot) }.toSet()
+    data class Rules(val shadowRules: RBAC.Builder, val actualRules: RBAC.Builder)
 
-            if (principals.isNotEmpty()) {
-                val policyName = clientsWithSelectors.joinToString(",") { it.compositeName() }
-                val policy: Policy.Builder = clientToPolicyBuilder.computeIfAbsent(policyName) {
-                    Policy.newBuilder().addAllPrincipals(principals)
-                }
+    private fun getRules(
+        serviceName: String,
+        incomingPermissions: Incoming,
+        snapshot: GlobalSnapshot,
+        roles: List<Role>
+    ): Rules {
 
-                val combinedPermissions = createCombinedPermissions(incomingEndpoint)
-                policy.addPermissions(combinedPermissions)
-            }
-        }
+        val incomingEndpointsPolicies = getIncomingEndpointPolicies(
+                serviceName, incomingPermissions, snapshot, roles
+        )
+
+        val restrictedEndpointsPolicies = incomingEndpointsPolicies.asSequence()
+            .filter { it.endpoint.unlistedClientsPolicy == IncomingEndpoint.UnlistedClientsPolicy.BLOCKANDLOG }
+            .map { (endpoint, policy) -> "$endpoint" to policy.build() }.toMap()
```

Comment:
I'm not sure if making a map key a `toString` of the whole object is a good idea. Won't it clutter config? Shouldn't it be something more predictable? I think we'll forget about it and unintentionally break something or make it even bigger by adding more fields to `IncomingEndpoint`.

I also think that `endpoint.toString()` is more readable than `"$endpoint"`. To me, it looks like some error.

Reply:
We have been wondering about it also. We decided to keep it that way because:
1. We need a unique policy name for every endpoint. ToString on an endpoint it's an easiest way to do it. Otherwise we would need to use some index in the policy name.
2. Policy name generated by toString looks good IMO. It has all the details of the endpoint which is based on, it may be handy for a debugging:
```
          "policies": {
            "IncomingEndpoint(path=/example, pathMatchingType=PATH, methods=[GET], clients=[ClientWithSelector(name=role1, selector=null)], unlistedClientsPolicy=BLOCKANDLOG)": {
              "permissions": [
                {
```

3. We have tests in `RBACFilterFactoryTest` which check what the policy name is. If it change unexpectedly we will see it. Every change in policy name has to be manually confirmed in the tests.

Reply:
For me `"$endpoint"` looks better than `endpoint.toString()`. I wonder what others think about it. LMK

Reply:
"$endpoint" is fine for me.